File: DAY_14/Devraj/Iterator_for_combination.cpp
class CombinationIterator {
public:
    string s;
    bool combs[15];
    int k;
    bool hasMoreCombs = true;
    void advanceCombs() {
        //move the first availabile and rightmost 1 left one step, sort from there
        for (int i = s.size() - 1; i; i--) {
            if (!combs[i] && combs[i - 1]) {
                swap(combs[i], combs[i - 1]);
                sort(begin(combs) + i + 1, begin(combs) + s.size(), greater<>());
                return;
            }
        }
        hasMoreCombs = false;
    }
    
    CombinationIterator(string &characters, int combinationLength) {
        s = characters;
        k = combinationLength;
        // initialising combs for our very first string
        for (int i = 0; i < 15; i++) combs[i] = i <  k;
    }
    
    string next() {
        // generating the current next string
        string res;
        for (int i = 0; i < s.size(); i++) if (combs[i]) res.push_back(s[i]);
        // preparing the next combinations
        advanceCombs();
        return res;
    }
    
    bool hasNext() {
     
        return hasMoreCombs;
    }
};

File: DAY_14/Devraj/Ravi/Iterator for Combination.java
class CombinationIterator {

    List<String> li = new ArrayList<>();
    int len;
    int ind[];
    String s;
    boolean hasRemItems = true;
    
    public CombinationIterator(String characters, int combinationLength) {
        len = combinationLength;
        s = characters;
        ind = new int[combinationLength];
        for (int i = 0; i < len; i++)
            ind[i] = i;
    }
    
    public String next() {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < ind.length; i++) {
            sb.append(s.charAt(ind[i]));
        }
        int j = 0;
        for (; j < len; j++) {
            if (ind[len-1-j] != s.length()-1-j) {
                ind[len-1-j]++;
                for (int k = len-j; k < len; k++) {
                    ind[k] = ind[k-1]+1;
                }
                break;
            }
        }
      
        if (j == len) hasRemItems = false;
        return sb.toString();
    }
    
    public boolean hasNext() {
        return hasRemItems;
    }
}
